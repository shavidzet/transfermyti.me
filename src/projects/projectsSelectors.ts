import { createSelector } from "reselect";
import * as R from "ramda";
import { selectIdToLinkedId } from "~/redux/reduxUtils";
import {
  areExistsInTargetShownSelector,
  toolNameByMappingSelector,
} from "~/allEntities/allEntitiesSelectors";
import { sourceTimeEntryCountByIdFieldSelectorFactory } from "~/timeEntries/timeEntriesSelectors";
import { activeWorkspaceIdSelector } from "~/workspaces/workspacesSelectors";
import {
  ReduxState,
  ProjectModel,
  ProjectsByIdModel,
  TableViewModel,
} from "~/typeDefs";

export const sourceProjectsByIdSelector = createSelector(
  (state: ReduxState) => state.projects.source,
  (sourceProjectsById): ProjectsByIdModel => sourceProjectsById,
);

export const sourceProjectsSelector = createSelector(
  sourceProjectsByIdSelector,
  (sourceProjectsById): ProjectModel[] => Object.values(sourceProjectsById),
);

export const targetProjectsByIdSelector = createSelector(
  (state: ReduxState) => state.projects.target,
  (targetProjectsById): ProjectsByIdModel => targetProjectsById,
);

const targetProjectsSelector = createSelector(
  targetProjectsByIdSelector,
  (targetProjectsById): ProjectModel[] => Object.values(targetProjectsById),
);

export const includedSourceProjectsSelector = createSelector(
  sourceProjectsSelector,
  (sourceProjects): ProjectModel[] =>
    sourceProjects.filter(sourceProject => sourceProject.isIncluded),
);

export const includedSourceProjectIdsSelector = createSelector(
  includedSourceProjectsSelector,
  (sourceProjects): string[] => R.pluck("id", sourceProjects),
);

export const sourceProjectsForTransferSelector = createSelector(
  includedSourceProjectsSelector,
  (sourceProjects): ProjectModel[] =>
    sourceProjects.filter(sourceProject => R.isNil(sourceProject.linkedId)),
);

export const sourceProjectsInActiveWorkspaceSelector = createSelector(
  sourceProjectsSelector,
  activeWorkspaceIdSelector,
  (sourceProjects, activeWorkspaceId) =>
    sourceProjects.filter(
      sourceProject => sourceProject.workspaceId === activeWorkspaceId,
    ),
);

export const projectIdToLinkedIdSelector = createSelector(
  sourceProjectsByIdSelector,
  (sourceProjectsById): Record<string, string> =>
    selectIdToLinkedId(sourceProjectsById),
);

export const projectsForInclusionsTableSelector = createSelector(
  areExistsInTargetShownSelector,
  sourceProjectsInActiveWorkspaceSelector,
  targetProjectsByIdSelector,
  sourceTimeEntryCountByIdFieldSelectorFactory("projectId"),
  (
    areExistsInTargetShown,
    sourceProjects,
    targetProjectsById,
    timeEntryCountByProjectId,
  ): TableViewModel<ProjectModel>[] =>
    sourceProjects.reduce((acc, sourceProject) => {
      const existsInTarget = sourceProject.linkedId !== null;
      if (existsInTarget && !areExistsInTargetShown) {
        return acc;
      }

      let isActiveInTarget = false;
      if (existsInTarget) {
        const targetId = sourceProject.linkedId as string;
        isActiveInTarget = targetProjectsById[targetId].isActive;
      }

      const entryCount = R.propOr<number, Record<string, number>, number>(
        0,
        sourceProject.id,
        timeEntryCountByProjectId,
      );

      return [
        ...acc,
        {
          ...sourceProject,
          entryCount,
          existsInTarget,
          isActiveInSource: sourceProject.isActive,
          isActiveInTarget,
        },
      ];
    }, [] as TableViewModel<ProjectModel>[]),
);

export const projectsTotalCountsByTypeSelector = createSelector(
  projectsForInclusionsTableSelector,
  projectsForInclusionsTable =>
    projectsForInclusionsTable.reduce(
      (
        acc,
        {
          entryCount,
          existsInTarget,
          isIncluded,
          isActiveInSource,
          isActiveInTarget,
        }: TableViewModel<ProjectModel>,
      ) => ({
        entryCount: acc.entryCount + entryCount,
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        isActiveInSource: acc.isActiveInSource + (isActiveInSource ? 1 : 0),
        isActiveInTarget: acc.isActiveInTarget + (isActiveInTarget ? 1 : 0),
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        entryCount: 0,
        existsInTarget: 0,
        isActiveInSource: 0,
        isActiveInTarget: 0,
        isIncluded: 0,
      },
    ),
);

export const projectsByWorkspaceIdByToolNameSelector = createSelector(
  toolNameByMappingSelector,
  sourceProjectsSelector,
  targetProjectsSelector,
  (toolNameByMapping, sourceProjects, targetProjects) => ({
    [toolNameByMapping.source]: groupProjectsByWorkspaceId(sourceProjects),
    [toolNameByMapping.target]: groupProjectsByWorkspaceId(targetProjects),
  }),
);

function groupProjectsByWorkspaceId(
  projects: ProjectModel[],
): Record<string, ProjectModel[]> {
  const projectsByWorkspaceId: Record<string, ProjectModel[]> = {};

  for (const project of projects) {
    const workspaceProjects = R.propOr<
      ProjectModel[],
      Record<string, ProjectModel[]>,
      ProjectModel[]
    >([], project.workspaceId, projectsByWorkspaceId);

    projectsByWorkspaceId[project.workspaceId] = [
      ...workspaceProjects,
      project,
    ];
  }

  return projectsByWorkspaceId;
}
