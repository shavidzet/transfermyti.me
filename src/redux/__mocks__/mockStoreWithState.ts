import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
