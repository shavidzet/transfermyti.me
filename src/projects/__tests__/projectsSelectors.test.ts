import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as projectsSelectors from "../projectsSelectors";

const TEST_STATE = {
  ...state,
  projects: {
    source: {
      "2001": {
        id: "2001",
        name: "Test Project A",
        workspaceId: "1001",
        clientId: "3001",
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#ea468d",
        estimate: { estimate: 0, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 3,
        linkedId: "clock-project-01",
        isIncluded: false,
        memberOf: "projects",
      },
      "2002": {
        id: "2002",
        name: "Test Project B",
        workspaceId: "1001",
        clientId: null,
        isBillable: false,
        isPublic: false,
        isActive: false,
        color: "#3750b5",
        estimate: { estimate: 40, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 3,
        linkedId: "clock-project-02",
        isIncluded: false,
        memberOf: "projects",
      },
      "2003": {
        id: "2003",
        name: "Test Project C",
        workspaceId: "1001",
        clientId: null,
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#c56bff",
        estimate: { estimate: 0, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 3,
        linkedId: "clock-project-03",
        isIncluded: false,
        memberOf: "projects",
      },
      "2004": {
        id: "2004",
        name: "Test Project D",
        workspaceId: "1001",
        clientId: null,
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#04bb9b",
        estimate: { estimate: 0, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 3,
        linkedId: null,
        isIncluded: true,
        memberOf: "projects",
      },
      "2005": {
        id: "2005",
        name: "Test Project E",
        workspaceId: "1001",
        clientId: null,
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#205500",
        estimate: { estimate: 0, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 4,
        linkedId: null,
        isIncluded: true,
        memberOf: "projects",
      },
      "2006": {
        id: "2006",
        name: "Test Project F",
        workspaceId: "1001",
        clientId: null,
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#a01aa5",
        estimate: { estimate: 0, type: "MANUAL" },
        userIds: ["6001"],
        entryCount: 4,
        linkedId: null,
        isIncluded: false,
        memberOf: "projects",
      },
    },
    target: {
      "clock-project-01": {
        id: "clock-project-01",
        name: "Test Project A",
        workspaceId: "clock-workspace-01",
        clientId: "clock-client-01",
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#ea468d",
        estimate: { estimate: "PT0S", type: "AUTO" },
        userIds: ["clock-user-01"],
        entryCount: 0,
        linkedId: "2001",
        isIncluded: false,
        memberOf: "projects",
      },
      "clock-project-02": {
        id: "clock-project-02",
        name: "Test Project B",
        workspaceId: "clock-workspace-01",
        clientId: "clock-client-02",
        isBillable: false,
        isPublic: false,
        isActive: true,
        color: "#c56bff",
        estimate: { estimate: "PT0S", type: "AUTO" },
        userIds: ["clock-user-01"],
        entryCount: 0,
        linkedId: "2002",
        isIncluded: false,
        memberOf: "projects",
      },
      "clock-project-03": {
        id: "clock-project-03",
        name: "Test Project C",
        workspaceId: "clock-workspace-01",
        clientId: "",
        isBillable: false,
        isPublic: false,
        isActive: false,
        color: "#04bb9b",
        estimate: { estimate: "PT0S", type: "AUTO" },
        userIds: ["clock-user-01"],
        entryCount: 0,
        linkedId: "2003",
        isIncluded: false,
        memberOf: "projects",
      },
    },
    isFetching: false,
  },
};

describe("within projectsSelectors", () => {
  cases(
    "selectors that directly access state return the correct value",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toEqual(options.expected);
    },
    [
      {
        name: "sourceProjectsByIdSelector returns state.sourceProjectsById",
        selector: projectsSelectors.sourceProjectsByIdSelector,
        expected: TEST_STATE.projects.source,
      },
      {
        name:
          "sourceProjectsSelector returns values from state.sourceProjectsById",
        selector: projectsSelectors.sourceProjectsSelector,
        expected: Object.values(TEST_STATE.projects.source),
      },
      {
        name: "targetProjectsByIdSelector returns state.targetProjectsById",
        selector: projectsSelectors.targetProjectsByIdSelector,
        expected: TEST_STATE.projects.target,
      },
    ],
  );

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceProjectsSelector",
        selector: projectsSelectors.includedSourceProjectsSelector,
      },
      {
        name: "for the includedSourceProjectIdsSelector",
        selector: projectsSelectors.includedSourceProjectIdsSelector,
      },
      {
        name: "for the sourceProjectsForTransferSelector",
        selector: projectsSelectors.sourceProjectsForTransferSelector,
      },
      {
        name: "for the sourceProjectsInActiveWorkspaceSelector",
        selector: projectsSelectors.sourceProjectsInActiveWorkspaceSelector,
      },
      {
        name: "for the projectIdToLinkedIdSelector",
        selector: projectsSelectors.projectIdToLinkedIdSelector,
      },
      {
        name: "for the projectsTotalCountsByTypeSelector",
        selector: projectsSelectors.projectsTotalCountsByTypeSelector,
      },
      {
        name: "for the projectsByWorkspaceIdByToolNameSelector",
        selector: projectsSelectors.projectsByWorkspaceIdByToolNameSelector,
      },
    ],
  );

  cases(
    "the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    options => {
      const updatedState = {
        ...TEST_STATE,
        allEntities: {
          ...TEST_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };
      const result = projectsSelectors.projectsForInclusionsTableSelector(
        updatedState,
      );

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );
});
