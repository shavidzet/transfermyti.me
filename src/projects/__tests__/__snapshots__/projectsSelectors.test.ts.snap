// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within projectsSelectors the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = false 1`] = `
Array [
  Object {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2004",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2006",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = true 1`] = `
Array [
  Object {
    "clientId": "3001",
    "color": "#ea468d",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2001",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-01",
    "memberOf": "projects",
    "name": "Test Project A",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#3750b5",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 40,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2002",
    "isActive": false,
    "isActiveInSource": false,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-02",
    "memberOf": "projects",
    "name": "Test Project B",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#c56bff",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2003",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-03",
    "memberOf": "projects",
    "name": "Test Project C",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2004",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2006",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors the selectors match their snapshots for the includedSourceProjectIdsSelector 1`] = `
Array [
  "2004",
  "2005",
]
`;

exports[`within projectsSelectors the selectors match their snapshots for the includedSourceProjectsSelector 1`] = `
Array [
  Object {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors the selectors match their snapshots for the projectIdToLinkedIdSelector 1`] = `
Object {
  "2001": "clock-project-01",
  "2002": "clock-project-02",
  "2003": "clock-project-03",
  "clock-project-01": "2001",
  "clock-project-02": "2002",
  "clock-project-03": "2003",
}
`;

exports[`within projectsSelectors the selectors match their snapshots for the projectsByWorkspaceIdByToolNameSelector 1`] = `
Object {
  "clockify": Object {
    "clock-workspace-01": Array [
      Object {
        "clientId": "clock-client-01",
        "color": "#ea468d",
        "entryCount": 0,
        "estimate": Object {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-01",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2001",
        "memberOf": "projects",
        "name": "Test Project A",
        "userIds": Array [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      Object {
        "clientId": "clock-client-02",
        "color": "#c56bff",
        "entryCount": 0,
        "estimate": Object {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-02",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2002",
        "memberOf": "projects",
        "name": "Test Project B",
        "userIds": Array [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      Object {
        "clientId": "",
        "color": "#04bb9b",
        "entryCount": 0,
        "estimate": Object {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-03",
        "isActive": false,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2003",
        "memberOf": "projects",
        "name": "Test Project C",
        "userIds": Array [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
    ],
  },
  "toggl": Object {
    "1001": Array [
      Object {
        "clientId": "3001",
        "color": "#ea468d",
        "entryCount": 3,
        "estimate": Object {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2001",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-01",
        "memberOf": "projects",
        "name": "Test Project A",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
      Object {
        "clientId": null,
        "color": "#3750b5",
        "entryCount": 3,
        "estimate": Object {
          "estimate": 40,
          "type": "MANUAL",
        },
        "id": "2002",
        "isActive": false,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-02",
        "memberOf": "projects",
        "name": "Test Project B",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
      Object {
        "clientId": null,
        "color": "#c56bff",
        "entryCount": 3,
        "estimate": Object {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2003",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-03",
        "memberOf": "projects",
        "name": "Test Project C",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
      Object {
        "clientId": null,
        "color": "#04bb9b",
        "entryCount": 3,
        "estimate": Object {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2004",
        "isActive": true,
        "isBillable": false,
        "isIncluded": true,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project D",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
      Object {
        "clientId": null,
        "color": "#205500",
        "entryCount": 4,
        "estimate": Object {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2005",
        "isActive": true,
        "isBillable": false,
        "isIncluded": true,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project E",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
      Object {
        "clientId": null,
        "color": "#a01aa5",
        "entryCount": 4,
        "estimate": Object {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2006",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project F",
        "userIds": Array [
          "6001",
        ],
        "workspaceId": "1001",
      },
    ],
  },
}
`;

exports[`within projectsSelectors the selectors match their snapshots for the projectsTotalCountsByTypeSelector 1`] = `
Object {
  "entryCount": 20,
  "existsInTarget": 3,
  "isActiveInSource": 5,
  "isActiveInTarget": 2,
  "isIncluded": 2,
}
`;

exports[`within projectsSelectors the selectors match their snapshots for the sourceProjectsForTransferSelector 1`] = `
Array [
  Object {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors the selectors match their snapshots for the sourceProjectsInActiveWorkspaceSelector 1`] = `
Array [
  Object {
    "clientId": "3001",
    "color": "#ea468d",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2001",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-01",
    "memberOf": "projects",
    "name": "Test Project A",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#3750b5",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 40,
      "type": "MANUAL",
    },
    "id": "2002",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-02",
    "memberOf": "projects",
    "name": "Test Project B",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#c56bff",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2003",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-03",
    "memberOf": "projects",
    "name": "Test Project C",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
  Object {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": Object {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2006",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;
